### PR TITLE
typing ir.py - Disallow untyped defs for ir.py

### PR DESCRIPTION
**Pull request by aorenste**
- Remove "mypy: allow-untyped-defs" and mark functions individually with "no-untyped-def"
- Mark some trivial functions with the proper return types (`None` and `torch.dtype`)
- Fixed a type bug in the signature of supported_dtype_of_cpp_wrapper()
- `ruff check torch/_inductor/ir.py --select ANN --fix --unsafe-fixes` and then fixed up things that looked incorrectly applied.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #131846
* __->__ #139238

cc @ezyang @malfet @xuzhao9 @gramster @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -831,20 +833,20 @@ def store_reduction(self, output_name, indexer, vars, reduction_vars):
         )
         return ops.store_reduction(output_name, indexer(vars), value)
 
-    def index_length(self):
+    def index_length(self):  # type: ignore[no-untyped-def]
```
Comment (Skylion007) [inline torch/_inductor/ir.py:836]:
```suggestion
    def index_length(self) -> int:
```

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -1,4 +1,3 @@
-# mypy: allow-untyped-defs
```
Comment (Skylion007) [inline torch/_inductor/ir.py:1]: Use ruff ANN rules to infer trivial return type annotations here

**Review by Skylion007 — APPROVED** [review]

**Review by ezyang — APPROVED** [review]
The "add per line suppresses" is not a very common strategy in our codebase but I guess it works

**Comment by aorenste** [comment]
> The "add per line suppresses" is not a very common strategy in our codebase but I guess it works

I have a stack of PRs which I'm slowly landing in the background which removes a lot of these per-line suppressions.  My general philosophy is to set the flags to prevent new missing types from being accidentally landed and then backfill the existing missing types.

**Comment by ezyang** [comment]
The main problem is that it is difficult to distinguish intentional suppressions from

**Comment by Skylion007** [comment]
> The main problem is that it is difficult to distinguish intentional suppressions from automatically added ones. This is less of a problem if you immediately fix them

We can add some other pragma that these are autogenerated, it only has to have type:ignore in the regex if I recall.

**Comment by aorenste** [comment]
For my info: what's an example or case where we add an automatically added suppression? Are these added and expected to be permanent additions or is it more like "some PR broke lint and we're fixing it forward but you should fix this"?

**Review by malfet — APPROVED** [review]

File: torch/_inductor/ir.py [hunk 3]
```diff
@@ -6672,11 +6675,11 @@ def should_realize_on_reuse(self, users):
             )
         return False
 
-    def mark_reuse(self, users):
+    def mark_reuse(self, users) -> None:  # type: ignore[no-untyped-def]
```
Comment (malfet) [inline torch/_inductor/ir.py:6678]:
```suggestion
    def mark_reuse(self, users) -> None:
```

File: torch/_inductor/ir.py [hunk 4]
```diff
@@ -6648,13 +6651,13 @@ def realize_hint(self):
...
 
-    def has_exceeded_max_reads(self):
+    def has_exceeded_max_reads(self):  # type: ignore[no-untyped-def]
         return isinstance(self.data, Pointwise) and (
             self.num_reads() > config.realize_acc_reads_threshold
             or self.has_large_inner_fn()
         )
 
-    def should_realize_on_reuse(self, users):
+    def should_realize_on_reuse(self, users):  # type: ignore[no-untyped-def]
```
Comment (malfet) [inline torch/_inductor/ir.py:6660]:
```suggestion
    def should_realize_on_reuse(self, users) -> bool:
```

File: torch/_inductor/ir.py [hunk 5]
```diff
@@ -6648,13 +6651,13 @@ def realize_hint(self):
         ):
             self.realize()
 
-    def has_exceeded_max_reads(self):
+    def has_exceeded_max_reads(self):  # type: ignore[no-untyped-def]
```
Comment (malfet) [inline torch/_inductor/ir.py:6654]:
```suggestion
    def has_exceeded_max_reads(self) -> bool:
```

File: torch/_inductor/ir.py [hunk 6]
```diff
@@ -6771,7 +6774,7 @@ def handle_sym_expr(stride):
         invoke_subgraph.outputs = outputs
         return outputs
 
-    def codegen(self, wrapper):
+    def codegen(self, wrapper) -> None:  # type: ignore[no-untyped-def]
```
Comment (malfet) [inline torch/_inductor/ir.py:6777]:
```suggestion
    def codegen(self, wrapper) -> None:
```

**Review by malfet — CHANGES_REQUESTED** [review]
You have an anti pattern of
```
def foo() -> None:  # type: ignore[no-untyped-def]
```
Please fix it (suggested a few changes)
Also, if you want to enforce it, you'll need to modify lint runner.yaml, don't you?

**Comment by aorenste** [comment]
> You have an anti pattern of
> 
> ```
> def foo() -> None:  # type: ignore[no-untyped-def]
> ```
> 
> Please fix it (suggested a few changes) Also, if you want to enforce it, you'll need to modify lint runner.yaml, don't you?

I think you're confused.  The purpose of this is to get us to a better state:
1. Disallow untyped defs by default and mark broken files individually (this was done in #127838)
2. When fixing a given file start by removing the file-level ignore and replace with function-level ignores (this PR)
3. Fix individual functions removing their function-level ignores. This is done in the rest of this stack (I did this during the last BE week but haven't had a lot of time to land) - this is the stack ending with #131852

The problem is that any one change can cause a ripple effect affecting MANY files so it's a lot easier to land smaller changes with temporary ignores than trying to land an single large change.

**Pull request closed without merging**